File: kombi-lib/src/main/java/com/sgnatiuk/combination/BinaryMaskIterator.java
package com.sgnatiuk.combination;

import java.util.Iterator;

class BinaryMaskIterator<T> implements Iterator<T> {

    private final CollectionBuilder<T> collectionBuilder;

    private final long lastValueMask;
    private long currentCombinationMask;

    public BinaryMaskIterator(Range range, CollectionBuilder<T> collectionBuilder) {
        this.collectionBuilder = collectionBuilder;

        currentCombinationMask = range.getFirst();
        lastValueMask = range.getLast();
    }

    @Override
    public boolean hasNext() {
        return currentCombinationMask <= lastValueMask;
    }

    @Override
    public T next() {
        T nextCollection = collectionBuilder.newCollection(Long.bitCount(currentCombinationMask));
        long itemsMask = currentCombinationMask;
        int index = 0;
        do {
            if ((itemsMask & 1) == 1) {
                collectionBuilder.addItemByIndex(nextCollection, index);
            }
            index++;
            itemsMask >>= 1;
        } while (itemsMask > 0);
        currentCombinationMask++;
        return nextCollection;
    }
}

File: kombi-lib/src/main/kotlin/com/sgnatiuk/combination/BinaryMaskIterator.kt
package com.sgnatiuk.combination

import com.sgnatiuk.extensions.bitCount

internal class BinaryMaskIterator<T>(
        combinationsRange: Range,
        private val collectionBuilder: CollectionBuilder<T>
) : MutableIterator<T> {

    private val lastValue: Long = combinationsRange.last

    private var combinationMask : Long = combinationsRange.first

    override fun hasNext() : Boolean {
        return combinationMask <= lastValue
    }

    override fun next(): T {
        val collection = collectionBuilder.newCollection(combinationMask.bitCount())
        var tempMask = combinationMask
        var index = 0
        do {
            if(tempMask and 1 == 1L) {
                collectionBuilder.addItemByIndex(collection, index)
            }
            index++
            tempMask = tempMask shr 1
        } while (tempMask > 0)
        combinationMask++
        return collection
    }

    override fun remove() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}
